# 4.6.2 Combining & Exporting Data

# This script contains the following points:

Import the data sets into Jupyter.
Check the dimensions of the imported dataframes.
Determine a suitable way to combine the orders_products_combined dataframe with the products dataframe.
Confirm the results of the merge using the merge flag.
Export the newly created dataframe as ords_prods_merge in a suitable format.

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [4]:
# 1. Import the data sets into Jupyter

In [16]:
## The main folder path
path= r'E:\Asawer\12.06.2024 Instacart Basket Analysis'

# Import dataset orders_products_combined.pkl
df_ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

# Import dataset products_checked.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

# 2. Check the dimensions of the imported dataframes

In [19]:
df_ords_prods_combined.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  order_dow  \
0             0           0   2539329        1             1          2   
1             0           0   2539329        1             1          2   
2             0           0   2539329        1             1          2   
3             0           0   2539329        1             1          2   
4             0           0   2539329        1             1          2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     7.0         196                  1   
1                  8                     7.0       14084                  2   
2                  8                     7.0       12427                  3   
3                  8                     7.0       26088                  4   
4                  8                     7.0       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [21]:
df_ords_prods_combined.shape

(32434489, 12)

In [23]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [25]:
df_prods.shape

(49672, 6)

# 3. Determine a suitable way to combine the orders_products_combined dataframe with the products dataframe

## NOTES:
we can see clearly from the procedures above that both of the dataframes have different shapes. 
A suitable way to combine them is to merge them using a default inner join on 'product_id'.

### Before combining the dataframes,I would prefer to drop unnecessary columns

In [30]:
### Drop unnamed columns in df_prods
df_prods = df_prods.drop('Unnamed: 0', axis=1)

In [32]:
df_prods.shape

(49672, 5)

In [34]:
### Drop unnamed columns in df_ords_combined
df_ords_prods_combined = df_ords_prods_combined.drop(columns = ['Unnamed: 0.1', 'Unnamed: 0', '_merge'])

In [36]:
df_ords_prods_combined.shape

(32434489, 9)

In [38]:
### Merge df_prods and df_ords_combined
df_merged = df_prods.merge(df_ords_prods_combined, on='product_id', indicator=True)

In [40]:
df_merged.shape

(32404859, 14)

In [ ]:
## 4. Confirm the results of the merge using the merge flag

In [44]:
df_merged['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

### Notes: 
After merging the two dataframes using inner join, the mew dataframe has 32,434,489 rows 
which means that only the observations found in both dataframes were merged. 

# 5. Export the newly created dataframe as ords_prods_merge in a suitable format

In [46]:
# Export data to pkl
df_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merged.pkl'))